### DISCUSSION TITLE
Elevation computation

### DISCUSSION THREAD

**Discussion by carlogico** [karttapullautin/karttapullautin#154] [Q&A] [answered]
Here: https://github.com/rphlo/karttapullautin/blob/9a6634dde7c67b9909a66b9a664be2617b07d0fb/src/contours.rs#L207-L216

As I understand it, this rounds the elevation in steps of `cinterval` (which is `scalefactor*0.3`), and if the difference between the original and the rounded value is less than .02 it changes the elevation to the rounded value ±.02. I cannot figure out why this is required, and what is the significance of the .02 (instead of something that depends on `cinterval`)?

**Comment by antbern** [comment]
I had the same thought while refactoring this. The only reason I could think of is that it makes sure that there is no height that is _exactly_ a multiple of the interval. It will always be either above or below. But if that is a useful property or not I'm not quite sure... :confused:

**Reply by carlogico** [reply]
That still doesn't seem like a convincing explanation: the value is modified even if it is not equal to the rounded value (`(*ele - temp).abs() < 0.02`). It seems that the elevation can be any continuous value except a set of width .04 centered around multiples of `cinterval`.

**Comment by rphlo** [comment]
Have you tried removing the block and checked the changes?
A bit of history:
Jarkko originaly wrote the code in BASIC for a random map generator that was then use for some early video game.
It was later translated to Perl and then adapted for use with actual Laser data map generation.
I then transcribed the code to Rust as closely as I could (not much comments in the Perl code).

There quite a lot of logic that were then only in Jarkko's mind and that might be lost in the translations...

**Reply by Jagge** [reply]
Thats true. When I was a kid I wrote basic program for Commodore 64 to print out random contours. It random generated DEM and calculated controus and printed them out to printer/paper. Later I ported it to PC just to not letting it all getting lost. Alpo K found somewhere and used it in his game. No much left from that era, but it sill is mentoned here: 
http://archive.orienteering.org.nz/newsletters/auckland/AOA_Apr_97_OCRO.pdf

![kuva](https://github.com/user-attachments/assets/1b6bbd5d-1c4c-44f9-8c04-923b6bbdbc81)

Then, again years later I thought it would be fun to generate maps for Catching Features game (it could import ocad8 files), so I converted the basic stuff to perl, just because it was mostly an easy search replace operation without having to think much at all. Again many years later National survey of Finland released lidar data as open data. I had no clue what that lidar was about, I just had heard one could use it for making O maps. I thought there is contours or something. Well, it was dissapointing, it was just lots of silly points But then at some point i figured I already have somewhere some old dirty code for processing data about like that, I would just need to remove the random part and replace it with reading lidar files in. So in no-time I got to spit out impressive looking maps that had some reseblance with reality or existing O maps. Somehow national map survey people found out my work and they invited me to give longish presentation at their lidar seminar where lidar pro people/companies were showcasing their pretty advanced systems. That was ridiculous, I guess they had no clue how clueless I was, just heard about lidar first time few weeks before, but hey, if you get an opportunity to do such a stand up comic gig of course you make it! Good times.

**Reply by carlogico** [reply]
Thanks for the context and history!

**Comment by Jagge** [comment, marked as answer]
If elevation grid point elavion equals with contour interval steps you will get contour topology issues (crossing/touching contours). This was implemented to avoid that.  0.02 (two centimeters) is just a random small number to avoid that issue, insignificant enough to matter, but big buffer enough to hopefully make it not get back to "bad value" for it getting rounded somewhere. Sure, it could be some fraction of contour interval, but in real world 2 cm is insignificant enough.
